core/common/src/annotation/TextBlock.ts

```diff
@@ -49,20 +62,48 @@ export interface TextBlockStringifyOptions {
    * Default: "undefined" - use "\t".
    */
   tabsAsSpaces?: number;
+  /** A string to insert in between the list marker and the list item.
+   * Default: " " - a single space..
+   */
+  listMarkerBreak?: string;
 }
 
-/** Abstract representation of any of the building blocks that make up a [[TextBlock]] document - namely [[Run]]s, [[Paragraph]]s, and [[TextBlock]] itself.
- * The [[TextBlock]] can specify an [AnnotationTextStyle]($backend) that formats its contents. Each component can specify an optional [[styleOverrides]] to customize that formatting.
+/**
+ * Context information that may be useful when converting a [[TextBlock]] to a string.
+ * @beta
+ */
+export interface TextBlockStringifyContext {
+  /** The current depth of the text block in the document structure. */
+  depth: number;
+}
+
+/**
+ * Abstract representation of any of the building blocks that make up a [[TextBlock]] document - namely [[Run]]s and [[ContainerBase]]s.
+ * The [[TextBlock]] can specify an [AnnotationTextStyle]($backend) that formats its contents.
+ * Each component can specify an optional [[styleOverrides]] to customize that formatting.
  * @beta
  */
 export abstract class TextBlockComponent {
   private _styleOverrides: TextStyleSettingsProps;
+  private _index: number = 0;
+
+  /** Discriminator field for the type of [[TextBlockComponent]]. */
+  public readonly abstract type: RunComponentType | ContainerComponentType;
```

Comment:
This is the root of all your casting issues. You have to define `type` only on the leaf types (Paragraph, List, TextRun, etc) so the compiler can use it to narrow types.